File: src/main/java/com/leijendary/spring/boot/template/error/NoHandlerFoundExceptionHandler.java
package com.leijendary.spring.boot.template.error;

import static com.leijendary.spring.boot.template.util.RequestContext.getLocale;
import static org.springframework.http.HttpStatus.NOT_FOUND;

import com.leijendary.spring.boot.template.data.response.ErrorResponse;

import org.springframework.context.MessageSource;
import org.springframework.core.annotation.Order;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.NoHandlerFoundException;

import lombok.RequiredArgsConstructor;

@RestControllerAdvice
@Order(1)
@RequiredArgsConstructor
public class NoHandlerFoundExceptionHandler {

    private final MessageSource messageSource;

    @ExceptionHandler(NoHandlerFoundException.class)
    @ResponseStatus(NOT_FOUND)
    public ErrorResponse catchNoHandlerFound(final NoHandlerFoundException exception) {
        final var code = "error.mapping.notFound";
        final var arguments = new Object[] { exception.getMessage() };
        final var message = messageSource.getMessage(code, arguments, getLocale());

        return ErrorResponse.builder()
                .addError("Request URL", code, message)
                .status(NOT_FOUND)
                .build();
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/error/NoHandlerFoundExceptionHandler.kt
package com.leijendary.spring.template.core.error

import com.leijendary.spring.template.core.data.ErrorData
import com.leijendary.spring.template.core.util.RequestContext.locale
import org.springframework.context.MessageSource
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.NoHandlerFoundException

@RestControllerAdvice
@Order(1)
class NoHandlerFoundExceptionHandler(private val messageSource: MessageSource) {
    @ExceptionHandler(NoHandlerFoundException::class)
    @ResponseStatus(NOT_FOUND)
    fun catchNoHandlerFound(exception: NoHandlerFoundException): List<ErrorData> {
        val code = "error.mapping.notFound"
        val arguments: Array<String?> = arrayOf(exception.message)
        val message = messageSource.getMessage(code, arguments, locale)
        val error = ErrorData(mutableListOf("request", "path"), code, message)

        return listOf(error)
    }
}
